pkg/ccl/changefeedccl/sink_kafka_v2.go

```diff
@@ -0,0 +1,612 @@
+}
+
+func buildKgoConfig(
+	ctx context.Context, u sinkURL, jsonStr changefeedbase.SinkSpecificJSONConfig,
+) ([]kgo.Opt, error) {
+	var opts []kgo.Opt
+
+	dialConfig, err := buildDialConfig(u)
+	if err != nil {
+		return nil, err
+	}
+
+	if dialConfig.tlsEnabled {
+		tlsCfg := &tls.Config{InsecureSkipVerify: dialConfig.tlsSkipVerify}
+		if dialConfig.caCert != nil {
+			caCertPool := x509.NewCertPool()
+			caCertPool.AppendCertsFromPEM(dialConfig.caCert)
+			tlsCfg.RootCAs = caCertPool
+		}
+
+		if dialConfig.clientCert != nil && dialConfig.clientKey == nil {
+			return nil, errors.Errorf(`%s requires %s to be set`, changefeedbase.SinkParamClientCert, changefeedbase.SinkParamClientKey)
+		} else if dialConfig.clientKey != nil && dialConfig.clientCert == nil {
+			return nil, errors.Errorf(`%s requires %s to be set`, changefeedbase.SinkParamClientKey, changefeedbase.SinkParamClientCert)
+		}
+
+		if dialConfig.clientCert != nil && dialConfig.clientKey != nil {
+			cert, err := tls.X509KeyPair(dialConfig.clientCert, dialConfig.clientKey)
+			if err != nil {
+				return nil, errors.Wrap(err, `invalid client certificate data provided`)
+			}
+			tlsCfg.Certificates = []tls.Certificate{cert}
+		}
+		opts = append(opts, kgo.DialTLSConfig(tlsCfg))
+	} else {
+		if dialConfig.caCert != nil {
+			return nil, errors.Errorf(`%s requires %s=true`, changefeedbase.SinkParamCACert, changefeedbase.SinkParamTLSEnabled)
+		}
+		if dialConfig.clientCert != nil {
+			return nil, errors.Errorf(`%s requires %s=true`, changefeedbase.SinkParamClientCert, changefeedbase.SinkParamTLSEnabled)
+		}
+	}
+
+	if dialConfig.saslEnabled {
+		var sasl sasl.Mechanism
+		switch dialConfig.saslMechanism {
+		case "OAUTHBEARER":
+			tp, err := newKgoOauthTokenProvider(ctx, dialConfig)
+			if err != nil {
+				return nil, err
+			}
+			sasl = sasloauth.Oauth(tp)
+		case "PLAIN", "":
+			sasl = saslplain.Plain(func(ctc context.Context) (saslplain.Auth, error) {
+				return saslplain.Auth{
+					User: dialConfig.saslUser,
+					Pass: dialConfig.saslPassword,
+				}, nil
+			})
+		case "SCRAM-SHA-256":
+			sasl = saslscram.Sha256(func(ctx context.Context) (saslscram.Auth, error) {
+				return saslscram.Auth{
+					User: dialConfig.saslUser,
+					Pass: dialConfig.saslPassword,
+				}, nil
+			})
+		case "SCRAM-SHA-512":
+			sasl = saslscram.Sha512(func(ctx context.Context) (saslscram.Auth, error) {
+				return saslscram.Auth{
+					User: dialConfig.saslUser,
+					Pass: dialConfig.saslPassword,
+				}, nil
+			})
+		default:
+			return nil, errors.Errorf(`unsupported SASL mechanism: %s`, dialConfig.saslMechanism)
+		}
+		opts = append(opts, kgo.SASL(sasl))
+	}
+
+	// Apply some statement level overrides. The flush related ones (Messages, MaxMessages, Bytes) are not applied here, but on the sinkBatchConfig instead.
+	// TODO: Remove the dependence on sarama eventually.
+	sinkCfg, err := getSaramaConfig(jsonStr)
```

Comment:
i think there's an issue with this approach -- is the user's intention with setting buffer size stuff to increase xput / decrease latency, or is it to protect kafka / align with their cluster settings. if it's the former, this is fine. if it's the latter, this will be ineffective, since this client is shared among the 8 parallelio workers, and so eg the batch size can be 8x what's configured.

i think the solution is to have one sink client per parallelio worker, and then the buffer size options will work for both use cases

Reply:
my workaround here is to use MaxMessages as the kgo limit. if the user specifies that as nonzero (it's 1k by default), it will be kgo's internal limit. we don't provide guarantees for the other Flush settings, only this one.